File: 0x4/bsearch/bsearch.go
// SPDX-License-Identifier: GPL-2.0
package bsearch

func BSearchInt(needle int, haystack []int) int {
	ret, _ := bSearchInt(needle, haystack)
	return ret
}

func BSearchIntWithTry(needle int, haystack []int) (int, int) {
	return bSearchInt(needle, haystack)
}

func bSearchInt(needle int, haystack []int) (int, int) {
	h := len(haystack) - 1
	l := 0

	var m int
	var try int
	for l <= h {
		try++
		m = (l + h) / 2
		switch {
		case needle < haystack[m]:
			h = m - 1
		case needle > haystack[m]:
			l = m + 1
		default:
			return m, try
		}
	}
	return -1, try
}

File: 0x4/bsearch.c
/* SPDX-License-Identifier: GPL-2.0 */

static int __bsearch(int v, const int a[], int n, int *try)
{
	int l, h;

	l = 0, h = n-1;
	while (l <= h) {
		int m = (l+h)/2;

		if (try)
			(*try)++;
		if (v < a[m])
			h = m-1;
		else if (v > a[m])
			l = m+1;
		else
			return m;
	}
	return -1;
}

int bsearch(int v, const int a[], int n)
{
	return __bsearch(v, a, n, 0);
}

int bsearch_with_try(int v, const int a[], int n, int *try)
{
	if (try)
		*try = 0;
	return __bsearch(v, a, n, try);
}
